larq_compute_engine/mlir/tests/prepare-tf.mlir

```diff
@@ -1,35 +1,41 @@
 // RUN: lce-tf-opt %s -tfl-prepare-lce | FileCheck %s --dump-input-on-failure
 
-func @bsign(%arg0: tensor<8x16xf32>) -> tensor<8x16xf32> {
+// CHECK-LABEL: @fuse_bsign
+func @fuse_bsign(%arg0: tensor<8x16xf32>) -> tensor<8x16xf32> {
   %0 = "tf.Sign"(%arg0) : (tensor<8x16xf32>) -> tensor<8x16xf32>
   %cst = constant dense<0.1> : tensor<f32>
   %2 = "tf.AddV2"(%0, %cst) : (tensor<8x16xf32>, tensor<f32>) -> tensor<8x16xf32>
   %3 = "tf.Sign"(%2) : (tensor<8x16xf32>) -> tensor<8x16xf32>
   return %3 : tensor<8x16xf32>
-// CHECK-LABEL: bsign
-// CHECK:  %0 = "tf.LqceBsign"(%arg0) : (tensor<8x16xf32>) -> tensor<8x16xf32>
-// CHECK:  return %0
+
+  // CHECK-NEXT: %0 = "tf.LqceBsign"(%arg0)
+  // CHECK-NEXT: return %0
 }
 
-func @bconv2d(%arg0: tensor<1x112x112x2xf32>) -> tensor<1x112x112x2xf32> {
+// CHECK-LABEL: @fuse_bconv2d
+func @fuse_bconv2d(%arg0: tensor<1x112x112x2xf32>) -> tensor<1x112x112x2xf32> {
   %cst = constant dense<[[[[1.0, -1.0], [1.0, 1.0]], [[-1.0, 1.0], [-1.0, 1.0]]]]> : tensor<1x2x2x2xf32>
   %0 = "tf.LqceBsign"(%arg0) : (tensor<1x112x112x2xf32>) -> tensor<1x112x112x2xf32>
   %1 = "tf.Conv2D"(%0, %cst) {padding = "SAME", strides = [1, 1, 1, 1]} : (tensor<1x112x112x2xf32>, tensor<1x2x2x2xf32>) -> tensor<1x112x112x2xf32>
   return %1 : tensor<1x112x112x2xf32>
-// CHECK-LABEL: bconv2d
-// CHECK:  %[[CST1:.*]] = constant dense<-2.000000e+00> : tensor<2xf32>
-// CHECK:  %[[CST2:.*]] = constant dense<4.000000e+00> : tensor<2xf32>
-// CHECK:  %[[TRP:.*]] = "tf.Transpose"
-// CHECK:  %[[CONV:.*]] = "tf.LqceBconv2d64"(%arg0, %[[TRP]], %[[CST1]], %[[CST2]]) {data_format = "NHWC", dilations = [1, 1, 1, 1], filter_format = "OHWI", padding = "SAME", strides = [1, 1, 1, 1]} : (tensor<1x112x112x2xf32>, tensor<2x1x2x2xf32>, tensor<2xf32>, tensor<2xf32>) -> tensor<1x112x112x2xf32>
-// CHECK:  return %[[CONV]]
+
+  // CHECK: %cst = constant
+  // CHECK: %[[fused_multiply:.*]] = constant dense<-2.000000e+00> : tensor<2xf32>
+  // CHECK: %[[fused_add:.*]] = constant dense<4.000000e+00> : tensor<2xf32>
+  // CHECK-NEXT: %[[transpose:.*]] = "tf.Transpose"(%cst
```

Comment:
What is the difference between `CHECK` and `CHECK-NEXT`?

Reply:
[`CHECK-NEXT`](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-next-directive) verifies that matches happen on exactly consecutive lines with no other lines in between them. So it is a bit stricter than just using `CHECK` here.

Reply:
From the docs that you linked to:

>  A “CHECK-NEXT:” cannot be the first directive in a file.

It looks like you should first do one `CHECK` and then use `CHECK-NEXT` for the following ones. The tests in `optimize.mlir` seem to start with `CHECK-NEXT`.

Reply:
I think this is fine since `CHECK-LABEL` is the first directive on any of the tests. E.g. see [this tflite test](https://github.com/tensorflow/tensorflow/blob/1d8b9fcff2951dc9b8eb0f66d3c1b45e7226e48d/tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir#L66-L73) which uses the same approach.

Reply:
Is there a closing bracket missing from this transpose line? Does the FileCheck system only do partial-matches or something and therefore it passes?

Reply:
This is intentional since we only want to partially match the first argument.